src/clr-angular/utils/smart-popover/providers/smart-popover-position.service.spec.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ *  Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ *
+ */
+
+import { Component, ElementRef } from '@angular/core';
+import { TestBed } from '@angular/core/testing';
+import { ClrAlignment } from '../enums/alignment.enum';
+import { ClrAxis } from '../enums/axis.enum';
+import { ClrSide } from '../enums/side.enum';
+import { ClrSmartPosition } from '../interfaces/smart-position.interface';
+import { ClrSmartPopoverEventsService, ClrSmartPopoverPositionService } from '@clr/angular';
+import { SMART_POPOVER_SERVICES } from '../smart-popover.module';
+
+@Component({
+  selector: 'test-host',
+  template: '',
+  providers: [SMART_POPOVER_SERVICES],
+})
+class TestHost {}
+
+interface TestContext {
+  eventService: ClrSmartPopoverEventsService;
+  positionService: ClrSmartPopoverPositionService;
+}
+
+export default function(): void {
+  describe('ClrSmartPopoverPositionService', function() {
+    describe('API', () => {
+      beforeEach(function(this: TestContext) {
+        TestBed.configureTestingModule({ declarations: [TestHost], providers: [SMART_POPOVER_SERVICES] });
+        const fixture = TestBed.createComponent(TestHost);
+        this.eventService = fixture.debugElement.injector.get(ClrSmartPopoverEventsService, null);
+        this.positionService = new ClrSmartPopoverPositionService(this.eventService, 'browser');
+      });
+
+      it('sets a position for the popover complex', function(this: TestContext) {
+        const testPosition: ClrSmartPosition = {
+          anchor: null,
+          content: null,
+          side: ClrSide.AFTER,
+          axis: ClrAxis.HORIZONTAL,
+        };
+        expect(this.positionService.position).toBeUndefined();
+        this.positionService.position = testPosition;
+        expect(this.positionService.position).toBe(testPosition);
+      });
+      it('aligns content when visibility is 🆗', function(this: TestContext) {
+        // TODO(matt): Figure out the correct tests here.
+        // the align, testVisibility and transformation functions are already tested in the operators
+        // I think I could flesh out the checking for each switch but this seems like it would duplicate much of the
```

Comment:
We can discuss this again, but due to the complexity of the service logic we need a sane way to test.

I think what you really need here is more of a test to check the if/else conditionals since operators does the deeper dive on specifics. However, you have to know this service is calling the right operators.